examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+ * if response is not in cache, the request fails).
+ *
+ * <p>This interceptor respects the cache-control directives in the server's response: max-age
+ * determines when the cache entry goes stale. no-cache, no-store, and no-transform entirely skip
+ * caching of the response. must-revalidate is ignored, as the cache does not support returning
+ * stale responses.
+ *
+ * <p>Note: other response headers besides cache-control (such as Expiration, Varies) are ignored by
+ * this implementation.
+ */
+final class SafeMethodCachingInterceptor implements ClientInterceptor {
+  public static CallOptions.Key<Boolean> NO_CACHE_CALL_OPTION =
+      CallOptions.Key.of("no-cache", false);
+  public static CallOptions.Key<Boolean> ONLY_IF_CACHED_CALL_OPTION =
+      CallOptions.Key.of("only-if-cached", false);
+
+  public static final class Key {
+    private final String fullMethodName;
+    private final MessageLite request;
+
+    public Key(String fullMethodName, MessageLite request) {
+      this.fullMethodName = fullMethodName;
+      this.request = request;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object instanceof Key) {
+        Key other = (Key) object;
+        return Objects.equals(this.fullMethodName, other.fullMethodName)
+            && Objects.equals(this.request, other.request);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(fullMethodName, request);
+    }
+  }
+
+  public static final class Value {
+    private final MessageLite response;
+    private final Deadline maxAgeDeadline;
+
+    public Value(MessageLite response, Deadline maxAgeDeadline) {
+      this.response = response;
+      this.maxAgeDeadline = maxAgeDeadline;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object instanceof Value) {
+        Value other = (Value) object;
+        return Objects.equals(this.response, other.response)
+            && Objects.equals(this.maxAgeDeadline, other.maxAgeDeadline);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(response, maxAgeDeadline);
+    }
+  }
+
+  public interface Cache {
+    void put(Key key, Value value);
+
+    Value get(Key key);
+
+    void remove(Key key);
+
+    void clear();
+  }
+
+  /**
+   * Obtain a new cache with a least-recently used eviction policy and the specified size limit. The
+   * backing caching implementation is provided by {@link LruCache}. It is safe for a single cache
+   * to be shared across multiple {@link SafeMethodCachingInterceptor}s without synchronization.
+   */
+  public static Cache newLruCache(final int cacheSizeInBytes) {
+    return new Cache() {
+      private final LruCache<Key, Value> lruCache =
+          new LruCache<Key, Value>(cacheSizeInBytes) {
+            protected int sizeOf(Key key, Value value) {
+              return value.response.getSerializedSize();
```

Comment:
micronit: the size from this method doesn't always match the size of the response.  This is particularly true for protos made with proto2 encoders, but parsed with proto3 definitions.  

I don't this its too damaging, but its going to be very subtly off.

Reply:
Acknowledged. I think it's impact is minor here